src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs

```diff
@@ -1229,10 +1226,51 @@ IEnumerable<int> M()
         yield return i;
     }
 }");
-            comp.VerifyDiagnostics(
-                // (7,26): error CS8176: Iterators cannot have by-reference locals
-                //         ref readonly int x = ref (new int[1])[0];
-                Diagnostic(ErrorCode.ERR_BadIteratorLocalType, "x").WithLocation(7, 26));
+            comp.VerifyEmitDiagnostics();
+        }
+
+        [Fact]
+        public void RefReadonlyInIterator_ForEach()
+        {
+            var source = """
+                using System.Collections.Generic;
+                class C
+                {
+                    int[] arr = new int[2];
+                    IEnumerable<int> M()
+                    {
+                        ref readonly int[] x = ref arr;
+
+                        foreach (var i in x)
+                        {
+                            System.Console.Write(i);
+                            yield return 1;
+                        }
+                
+                        foreach (var j in x)
+                        {
+                            System.Console.Write(j);
+                            yield return 2;
+                        }
+                    }
+                }
+                """;
+
+            CreateCompilation(source, parseOptions: TestOptions.Regular12).VerifyDiagnostics(
+                // (7,28): error CS8652: The feature 'ref and unsafe in async and iterator methods' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         ref readonly int[] x = ref arr;
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "x").WithArguments("ref and unsafe in async and iterator methods").WithLocation(7, 28));
+
+            var expectedDiagnostics = new[]
+            {
+                // (15,27): error CS9217: A 'ref' local cannot be preserved across 'await' or 'yield' boundary.
```

Comment:
> // (15,27): error CS9217: A 'ref' local cannot be preserved across 'await' or 'yield' boundary. [](http://example.com/codeflow?start=15&length=96)

Why we are not reporting a similar error in ```foreach (var i in x)```? Is it really going to work? If so, consider adding a test that executes the code and observes the result.

Reply:
Yes, it works. The ref local is not used across yield, it's used only once at the beginning to get the array value for the foreach to operate on. I will add a test.